src/inference/src/ie_core.cpp

```diff
@@ -971,125 +988,138 @@ class CoreImpl : public ie::ICore, public std::enable_shared_from_this<ie::ICore
     ov::InferencePlugin GetCPPPluginByName(const std::string& pluginName) const {
         OV_ITT_SCOPE(FIRST_INFERENCE, ie::itt::domains::IE_LT, "CoreImpl::GetCPPPluginByName");
 
-        std::lock_guard<std::mutex> lock(pluginsMutex);
         auto deviceName = pluginName;
         if (deviceName == ov::DEFAULT_DEVICE_NAME)
             deviceName = "AUTO";
         stripDeviceName(deviceName, "-");
-        auto it = pluginRegistry.find(deviceName);
-        if (it == pluginRegistry.end()) {
-            if (pluginName == ov::DEFAULT_DEVICE_NAME)
-                IE_THROW() << "No device is provided, so AUTO device is used by default, which failed loading.";
-            else
-                IE_THROW() << "Device with \"" << deviceName << "\" name is not registered in the InferenceEngine";
+        std::map<std::string, PluginDescriptor>::const_iterator it;
+        {
+            // Global lock to find plugin.
+            // Always use global mutex if iterate over plugins or pluginRegistry
+            std::lock_guard<std::mutex> g_lock(get_mutex());
+
+            // Plugin is not created, check that plugin is registered
+            it = pluginRegistry.find(deviceName);
+            if (it == pluginRegistry.end()) {
+                if (pluginName == ov::DEFAULT_DEVICE_NAME)
+                    IE_THROW() << "No device is provided, so AUTO device is used by default, which failed loading.";
+                else
+                    IE_THROW() << "Device with \"" << deviceName << "\" name is not registered in the InferenceEngine";
+            }
         }
+        std::lock_guard<std::mutex> lock(get_mutex(deviceName));
 
+        PluginDescriptor desc;
+        {
+            // Global lock to find plugin.
+            // Always use global mutex if iterate over plugins or pluginRegistry
+            std::lock_guard<std::mutex> g_lock(get_mutex());
```

Comment:
here we are trying to lock second mutex while global one is already locked. Seems dangerous since we probably can have situation when on thread locked one mutex and waits for another, while another threads does the things in reverse order and waits as well.

Reply:
The logic is here, that we lock global mutex only for cases when we work with global factories. In this PR I also changed the logic in order to decrease sections where we work with these factories.